index.js

```diff
@@ -4413,6 +4413,53 @@
     impl: matchAll
   };
 
+  //# replace :: (Array (Maybe String) -> String) -> RegExp -> String -> String
+  //.
+  //. Replaces occurrences of the given pattern within the given string
+  //. in accordance with the given replacement function, which receives an
+  //. array of captured values. Replaces all occurrences of the pattern if
+  //. its `g` flag is set; just the first occurrence otherwise.
+  //.
+  //. ```javascript
+  //. > S.replace (([$1]) => S.maybe ('') (S.toUpper) ($1)) (/(\w)/) ('foo')
+  //. 'Foo'
+  //.
+  //. > S.replace (([$1]) => S.maybe ('') (S.toUpper) ($1)) (/(\w)/g) ('foo')
+  //. 'FOO'
+  //.
+  //. > S.replace (S.show) (/(foo)(bar)?/) ('<>')
+  //. '<>'
+  //.
+  //. > S.replace (S.show) (/(foo)(bar)?/) ('<foo>')
+  //. '<[Just ("foo"), Nothing]>'
+  //.
+  //. > S.replace (S.show) (/(foo)(bar)?/) ('<foobar>')
+  //. '<[Just ("foo"), Just ("bar")]>'
+  //. ```
+  function replace(substitute) {
+    return function(pattern) {
+      return function(text) {
+        return text.replace (pattern, function() {
+          var groups = [];
+          var group, idx = 1;
+          //  eslint-disable-next-line no-plusplus
+          while (typeof (group = arguments[idx++]) !== 'number') {
+            groups.push (group == null ? Nothing : Just (group));
+          }
```

Comment:
Why do you normally follow the no-plusplus rule, and how is this an exception?

Reply:
I eschew the prefix increment operator (`++idx`) because it is a special case of `+=`, and I dislike special cases:

```console
$ python -c 'import this' | grep 'only one'
There should be one-- and preferably only one --obvious way to do it.
```

I use the postfix increment operator (`idx++`) only when it enables a clearer algorithm than is possible with `+=`. This is the case here. Admittedly, we could use `idx += 1` and change the initial value of `idx` from `1` to `0`, but it would then be less clear that the first element is skipped (and observant readers might wonder whether the first element is being skipped unintentionally).